compiler/resolution/functionResolution.cpp

```diff
@@ -5184,81 +5194,101 @@ resolveNew(CallExpr* call)
       // argument if we're making a call to an initializer
-      Type* typeToNew = toReplace->symbol()->typeInfo();
-      VarSymbol* new_temp = newTemp("new_temp", typeToNew);
+      Type*      typeToNew = toReplace->symbol()->typeInfo();
+      VarSymbol* newTmp    = newTemp("new_temp", typeToNew);
+
+      VarSymbol* newMT     = newTemp("_mt",      dtMethodToken);
+
       if (typeToNew->symbol->hasFlag(FLAG_GENERIC)) {
-        USR_FATAL(call, "Sorry, new style initializers don't work with generics yet.  Stay tuned!");
+        USR_FATAL(call,
+                  "Sorry, new style initializers don't work with "
+                  "generics yet.  Stay tuned!");
       }
-      DefExpr* def = new DefExpr(new_temp);
-      Expr* insertBeforeMe = ((!isBlockStmt(call->parentExpr)) ?
-                              call->parentExpr : call);
+
+      DefExpr* def            = new DefExpr(newTmp);
+      Expr*    insertBeforeMe = NULL;
+
+      if (isBlockStmt(call->parentExpr) == true) {
+        insertBeforeMe = call;
+
+      } else {
+        insertBeforeMe = call->parentExpr;
+      }
+
       insertBeforeMe->insertBefore(def);
+
       resolveExpr(def);
 
       if (!isRecord(typeToNew) && !isUnion(typeToNew)) {
-        BlockStmt* allocCallBlock = new BlockStmt();
+        BlockStmt* allocBlock = new BlockStmt();
+        CallExpr*  allocCall  = callChplHereAlloc(typeToNew);
+
+        insertBeforeMe->insertBefore(allocBlock);
 
-        CallExpr* innerCall = callChplHereAlloc(typeToNew);
-        CallExpr* allocCall = new CallExpr(PRIM_MOVE, new_temp,
-                                           innerCall);
-        insertBeforeMe->insertBefore(allocCallBlock);
-        allocCallBlock->insertAtTail(allocCall);
+        allocBlock->insertAtTail(new CallExpr(PRIM_MOVE,   newTmp, allocCall));
+        allocBlock->insertAtTail(new CallExpr(PRIM_SETCID, newTmp));
 
-        CallExpr* setIdCall = new CallExpr(PRIM_SETCID, new_temp);
-        allocCallBlock->insertAtTail(setIdCall);
-        normalize(allocCallBlock);
-        resolveBlockStmt(allocCallBlock);
+        normalize(allocBlock);
+
+        resolveBlockStmt(allocBlock);
       }
-      call->insertAtTail(new NamedExpr("meme", new SymExpr(new_temp)));
+
+      call->insertAtHead(new SymExpr(newTmp));
+      call->insertAtHead(new SymExpr(newMT));
     }
 
-    // Now finish resolving it, since we are after all in
-    // resolveNew.
     resolveExpr(call);
   }
 
   // Do some error checking
-  if (FnSymbol* fn = call->isResolved())
-  {
-    // If the function is a constructor, OK
-    if (fn->hasFlag(FLAG_CONSTRUCTOR))
-      return;
-    else // otherwise, issue an error
-      USR_FATAL(call, "invalid use of 'new' on %s", fn->name);
-  }
+  if (FnSymbol* fn = call->isResolved()) {
+    if (fn->hasFlag(FLAG_CONSTRUCTOR)) {
 
-  if (call->get(1)) {
-    Expr* arg = call->get(1);
+    } else if (fn->hasFlag(FLAG_METHOD) && strcmp(fn->name, "init") == 0) {
```

Comment:
I believe this branch will never be reached, as iirc "init" functions are also marked with FLAG_CONSTRUCTOR.  Unless I missed that change in here?